File: crates/flighthq-scene-gl/src/unlit_gl_mesh_material_renderer.rs
//! The built-in Unlit forward renderer.
//!
//! Ports `@flighthq/scene-gl` `unlitGlMeshMaterialRenderer.ts`.

use flighthq_materials::unlit_materials::unlit_material_kind;
use flighthq_render_gl::GlRenderState;
use flighthq_types::camera::Camera;
use flighthq_types::scene_render::{SceneLightBlock, SceneRenderProxy};

use crate::gl_mesh_material_registry::{
    GlMeshMaterialRenderer, MeshMaterial, register_gl_mesh_material_renderer,
};
use crate::gl_scene_runtime::{GlMeshUpload, GlSceneRuntime};

/// The built-in Unlit renderer.
pub struct UnlitGlMeshMaterialRenderer;

impl GlMeshMaterialRenderer for UnlitGlMeshMaterialRenderer {
    fn bind(
        &self,
        _state: &mut GlRenderState,
        _scene: &mut GlSceneRuntime,
        _material: Option<&dyn MeshMaterial>,
        _lights: &SceneLightBlock,
        _camera: &Camera,
    ) {
        // Stub: requires ensure_gl_unlit_program + begin_gl_mesh_draw +
        // set_gl_mesh_view_projection + bind_gl_unlit_surface.
    }

    fn draw(
        &self,
        _state: &mut GlRenderState,
        _scene: &mut GlSceneRuntime,
        _proxy: &SceneRenderProxy,
        _upload: &GlMeshUpload,
    ) {
        // Stub: requires draw_gl_mesh_subset.
    }
}

/// Registers the built-in Unlit renderer for `UnlitMaterialKind`.
pub fn register_unlit_gl_material(scene: &mut GlSceneRuntime) {
    register_gl_mesh_material_renderer(
        scene,
        unlit_material_kind(),
        Box::new(UnlitGlMeshMaterialRenderer),
    );
}

#[cfg(test)]
mod tests {
    use super::*;
    use flighthq_materials::unlit_materials::unlit_material_kind;

    use crate::gl_mesh_material_registry::get_gl_mesh_material_renderer;
    use crate::gl_scene_runtime::create_gl_scene_runtime;

    // register_unlit_gl_material

    #[test]
    fn register_unlit_gl_material_registers_for_the_unlit_kind() {
        let mut scene = create_gl_scene_runtime();
        register_unlit_gl_material(&mut scene);
        assert!(get_gl_mesh_material_renderer(&scene, unlit_material_kind()).is_some());
    }
}

File: packages/scene-gl/src/unlitGlMeshMaterialRenderer.ts
import { unpackColorToLinear } from '@flighthq/color';
import type {
  LinearColor,
  Camera3D,
  GlMeshMaterialRenderer,
  GlRenderState,
  Material,
  MeshGeometry,
  SceneLightBlock,
  SceneRenderProxy,
  UnlitMaterial,
  GlUnlitDefineKey,
} from '@flighthq/types';
import { UnlitMaterialKind } from '@flighthq/types';

import { registerGlMeshMaterialRenderer } from './glMeshMaterialRegistry';
import {
  beginGlMeshDraw,
  bindGlUvTransform,
  drawGlMeshSubset,
  hasGlUvTransform,
  setGlMeshViewProjection,
} from './glMeshProgram';
import { getGlSceneRuntime } from './glSceneRuntime';
import { bindGlUnlitSurface, ensureGlUnlitProgram } from './glUnlitPrelude';

// The built-in Unlit forward renderer (GlMeshMaterialRenderer for UnlitMaterialKind). Lighting-
// independent flat color: bind selects the unlit variant for the material's base-color map / alpha
// mode, uploads the camera view-projection and the linear base color, and draw issues the indexed
// draw. Lights are ignored. See registerUnlitGlMaterial to install it.
export const unlitGlMeshMaterialRenderer: GlMeshMaterialRenderer = {
  bind(
    state: GlRenderState,
    material: Readonly<Material> | null,
    _lights: Readonly<SceneLightBlock>,
    camera: Readonly<Camera3D>,
  ): void {
    const gl = state.gl;
    const unlit = material as Readonly<UnlitMaterial> | null;
    const program = ensureGlUnlitProgram(state, defineKeyForMaterial(unlit));
    beginGlMeshDraw(state, program, unlit !== null && unlit.doubleSided);
    setGlMeshViewProjection(gl, program.locViewProjection, camera);

    if (unlit === null) {
      bindGlUnlitSurface(state, program, WHITE, 1, null, 0.5);
      return;
    }
    unpackColorToLinear(scratchRgba, unlit.baseColor);
    bindGlUnlitSurface(state, program, scratchRgba, 1, unlit.baseColorMap, unlit.alphaCutoff);
    bindGlUvTransform(gl, program, unlit.baseColorMap);
  },

  draw(state: GlRenderState, proxy: Readonly<SceneRenderProxy>, geometry: Readonly<MeshGeometry>): void {
    const program = getGlSceneRuntime(state).activeMeshProgram;
    if (program === null) return;
    drawGlMeshSubset(state, program, proxy, geometry);
  },
};

// Registers the built-in Unlit renderer for UnlitMaterialKind on this state. Opt-in (no top-level
// side effect); call once per GlRenderState before drawScene so meshes with UnlitMaterials draw.
export function registerUnlitGlMaterial(state: GlRenderState): void {
  registerGlMeshMaterialRenderer(state, UnlitMaterialKind, unlitGlMeshMaterialRenderer);
}

function defineKeyForMaterial(material: Readonly<UnlitMaterial> | null): GlUnlitDefineKey {
  return {
    alphaMaskEnabled: material !== null && material.alphaMode === 'mask',
    hasColorMap: material !== null && material.baseColorMap !== null && material.baseColorMap.image !== null,
    hasUvTransform: hasGlUvTransform(material !== null ? material.baseColorMap : null),
    vertexColor: false,
  };
}

const scratchRgba: LinearColor = [0, 0, 0, 0];
const WHITE: LinearColor = [1, 1, 1, 1];
